src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,260 @@
+                                break;
+                            case { Condition: IsPatternExpressionSyntax { Expression: var patternInput, Pattern: ConstantPatternSyntax { Expression: var patternExpression } } }:
+                                left = patternInput;
+                                right = patternExpression;
+                                break;
+                            case { Condition: InvocationExpressionSyntax { Expression: var receiver, ArgumentList.Arguments: { Count: 2 } arguments } }
+                                when referenceEqualsMethod != null && referenceEqualsMethod.Equals(semanticModel.GetSymbolInfo(receiver, cancellationToken).Symbol):
+
+                                left = arguments[0].Expression;
+                                right = arguments[1].Expression;
+                                break;
+
+                            default:
+                                continue;
+                        }
+
+                        if (!AreOperandsApplicable(left, right, out parameter)
+                            && !AreOperandsApplicable(right, left, out parameter))
+                        {
+                            continue;
+                        }
+
+                        var throwStatement = ifStatement.Statement switch
+                        {
+                            ThrowStatementSyntax @throw => @throw,
+                            BlockSyntax { Statements: { Count: 1 } statements } => statements[0] as ThrowStatementSyntax,
+                            _ => null
+                        };
+
+                        if (throwStatement?.Expression is not ObjectCreationExpressionSyntax thrownInIf
+                            || !IsConstructorApplicable(thrownInIf, parameter))
+                        {
+                            continue;
+                        }
+
+                        break;
+
+                    // this.field = param ?? throw new ArgumentNullException(nameof(param));
+                    case ExpressionStatementSyntax
+                    {
+                        Expression: AssignmentExpressionSyntax
+                        {
+                            Right: BinaryExpressionSyntax
+                            {
+                                OperatorToken.RawKind: (int)SyntaxKind.QuestionQuestionToken,
+                                Left: ExpressionSyntax maybeParameter,
+                                Right: ThrowExpressionSyntax { Expression: ObjectCreationExpressionSyntax thrownInNullCoalescing }
+                            }
+                        }
+                    }:
+                        if (!IsParameter(maybeParameter, out parameter) || !IsConstructorApplicable(thrownInNullCoalescing, parameter))
+                        {
+                            continue;
+                        }
+
+                        break;
+
+                    default:
+                        continue;
+                }
```

Comment:
i recommend extracting this out to a helper that just returns the parameter.  the nested switch/break/continue is pretty hard to erady (for me at least).  tn you.

Reply:
Yes, this leans pretty hard on definite assignment to ensure correctness. I'm open to making the change you've described.

Reply:
i think you could still depend on it.  you'd just have two separate methods for the If vs Assign cases.